optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,422 @@
+                "The parameter distribution should be one of the {}".format(
+                    distribution, distribution_list
+                )
+            )
+
+        assert low < high
+
+        return low, high, q
+
+    def _transform_to_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
+            else:
+                distribution_list = [
+                    distributions.UniformDistribution.__name__,
+                    distributions.LogUniformDistribution.__name__,
+                    distributions.DiscreteUniformDistribution.__name__,
+                    distributions.IntUniformDistribution.__name__,
+                    distributions.IntLogUniformDistribution.__name__,
+                    distributions.CategoricalDistribution.__name__,
+                ]
+                raise NotImplementedError(
+                    "The distribution {} is not implemented. "
+                    "The parameter distribution should be one of the {}".format(
+                        distribution, distribution_list
+                    )
+                )
+        return transformed
+
+    def _transform_from_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.exp(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                q = self._q[param_name]
+                low = self._low[param_name]
+                high = self._high[param_name]
+                samples = (np.round((samples - low) / q - 0.5) + 0.5) * q + low
```

Comment:
This is quite a tricky part. 
Since `low = distribution.low - 0.5 * q`, this code is OK. 
`samples` are taken from `UnifromDistribution(distribution.low - 0.5*q, distribution.high + 0.5*q)`, argument of `np.round`, `(samples - low) / q - 0.5`, is included in [-0.5, (distribution.high - distribution.low) / q + 0.5].

For clarification, I will add some comments here.

Reply:
How about this?
```suggestion
                samples = np.round((samples - distribution.low) / q) * q + distribution.low
```

Reply:
Hmm... For example, please consider the case of `low =1.2, high = 2.4, q = 0.3`. This case the search space will be `{1.2, 1.5, 1.8, 2.1, 2.4}`. Let `sample = 1.7`. 

My suggested code results in `round((s - low) / q) * q + low = round((1.7 - 1.2) / 0.3) * 0.3 + 1.2 = 1.8`. This `1.8` is in the search space.

The current code results in `(round((1.7 - 1.2) / 0.3 - 0.5) + 0.5) * 0.3 + 1.2 = 1.65`. This `1.65` is not in the search space.

Reply:
I see. The `low` and `high` are not the original `low` and `high`. Yo new suggested code looks easy to understand. Great!